### PR TITLE
Fix issues that occured when running ppafm-generate-elff with tip's density

### LINKED ISSUE #312: FDBM (Full-density based method) broken by recent changes
The electrostatic and vdW calculations do not currently work in FDBM, as you can test e.g. with the `ppafm/examples/pyridineDensOverlap/` example.

Running `ppafm-generate-elff -i sample/LOCPOT.xsf --tip_dens tip/CHGCAR.xsf --Rcore 0.7 --doDensity` leads to the error message
``NameError: name 'parameters' is not defined`` 
issued inside the `ppafm/ppafm/HighLevel.py/getAtomsWhichTouchPBCcell` function when trying to call
``atoms, nDim, lvec = io.loadGeometry(fname, format=geometry_format, parameters=parameters)``

Similarly, calling `ppafm-generate-dftd3 -i sample/LOCPOT.xsf --df_name PBE` gives
``ValueError: Please provide the parameters dictionary here``,
raised inside the `loadGeometry` function itself, after this function has been called from `ppafm/HighLevel.py/computeDFTD3` by the line
``atoms, nDim, lvec = io.loadGeometry(input_file, format=geometry_format, parameters=parameters)``

### PR DESCRIPTION
**Pull request by yakutovicha**
fixes #312

blocked by [this](https://github.com/Probe-Particle/ppafm/issues/314#issuecomment-2528424010).

### CODE REVIEW

File: examples/pyridineDensOverlap/example_pyridine_density_overlap.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    if not Path("LOCPOT.xsf").exists():
+        urllib.request.urlretrieve("https://zenodo.org/records/14222456/files/pyridine.zip?download=1", "sample.zip")
+        # Unzip the file using python tools only
+        with zipfile.ZipFile("sample.zip", "r") as zip_ref:
+            zip_ref.extractall(".")
+
+    generate_conv_rho(["-s", "LOCPOT.xsf", "-t", "tip/CHGCAR.xsf", "-B", "1.0", "-E"])
+    generate_elff(["-i", "LOCPOT.xsf", "--tip_dens", "tip/CHGCAR.xsf", "--Rcore", "0.7", "--energy", "--doDensity"])
+    generate_dftd3(["--input", "LOCPOT.xsf", "--df_name", "PBE"])
```
Comment (yakutovicha) [inline examples/pyridineDensOverlap/example_pyridine_density_overlap.py:27]: Same issue as above pops up if I change "--input" back to "-i"

**Review by mondracek — APPROVED** [review]

**Comment by NikoOinonen** [comment]
I think I see where the segfault is coming from. The first two commands use the `--energy` option which sets the pointer for the energy array on the C++ side. The third command (dftd3) does not have the option, but the pointer still remains even though it might not be valid anymore. A stopgap fix is to simply add the `--energy` option also to the dftd3 command.

**Comment by NikoOinonen** [comment]
For the `-E` argument issue, change this line : https://github.com/Probe-Particle/ppafm/blob/12a90344f703da0278abe39bf067a81fafb95b70/ppafm/common.py#L323
into this:
```python
import copy
arg_dict = copy.deepcopy(self.cli_args[name])
```

**Comment by yakutovicha** [comment]
> I think I see where the segfault is coming from. The first two commands use the `--energy` option which sets the pointer for the energy array on the C++ side. The third command (dftd3) does not have the option, but the pointer still remains even though it might not be valid anymore. A stopgap fix is to simply add the `--energy` option also to the dftd3 command.

It's working! 👍

File: examples/pyridineDensOverlap/example_pyridine_density_overlap.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    os.chdir(script_location)
+
+    if not Path("LOCPOT.xsf").exists():
+        urllib.request.urlretrieve("https://zenodo.org/records/14222456/files/pyridine.zip?download=1", "sample.zip")
+        # Unzip the file using python tools only
+        with zipfile.ZipFile("sample.zip", "r") as zip_ref:
+            zip_ref.extractall(".")
+
+    generate_conv_rho(["-s", "LOCPOT.xsf", "-t", "tip/CHGCAR.xsf", "-B", "1.0", "-E"])
+    generate_elff(["-i", "LOCPOT.xsf", "--tip_dens", "tip/CHGCAR.xsf", "--Rcore", "0.7", "--energy", "--doDensity"])
```
[the remark this answers is not in the document]
Comment (yakutovicha) [reply examples/pyridineDensOverlap/example_pyridine_density_overlap.py]: workaround proposed by @NikoOinonen: https://github.com/Probe-Particle/ppafm/pull/316#issuecomment-2531178177

File: examples/pyridineDensOverlap/example_pyridine_density_overlap.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        urllib.request.urlretrieve("https://zenodo.org/records/14222456/files/CO_tip.zip?download=1", "tip.zip")
+        # Unzip the file using python tools only
+        with zipfile.ZipFile("tip.zip", "r") as zip_ref:
+            # Make "tip" directory if it does not exist
+            Path("tip").mkdir(exist_ok=True)
+            zip_ref.extractall("tip")
+
+    generate_conv_rho("-s CHGCAR.xsf -t tip/density_CO.xsf -B 1.0 -E".split())
+    generate_elff("-i LOCPOT.xsf --tip_dens tip/density_CO.xsf --Rcore 0.7 -E --doDensity".split())
+    generate_dftd3("-i LOCPOT.xsf --df_name PBE".split())
```
Comment (NikoOinonen) [inline examples/pyridineDensOverlap/example_pyridine_density_overlap.py:35]: The `--energy` is missing again.

File: examples/pyridineDensOverlap/example_pyridine_density_overlap.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (NikoOinonen) [reply examples/pyridineDensOverlap/example_pyridine_density_overlap.py]: Resolved below.

**Review by NikoOinonen — APPROVED** [review]

File: examples/pyridineDensOverlap/example_pyridine_density_overlap.py (hunk 3 above)
Comment (yakutovicha) [reply examples/pyridineDensOverlap/example_pyridine_density_overlap.py]: fixed in https://github.com/Probe-Particle/ppafm/pull/316/commits/440ef7c26f2008147aa7f5ed8e1092d43759cc5e

File: examples/pyridineDensOverlap/example_pyridine_density_overlap.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yakutovicha) [reply examples/pyridineDensOverlap/example_pyridine_density_overlap.py]: Solved in https://github.com/Probe-Particle/ppafm/pull/316/commits/7c9806f7a8a4cb732ed1cc609c05e51d3f7da6df following this suggestion: https://github.com/Probe-Particle/ppafm/pull/316#issuecomment-2531216466.

**Pull request merged**